app.py

```diff
@@ -804,19 +805,38 @@ def tree_process(tree : ParentedTree, senttok: List[str]) -> ParentedTree:
 			subt.label = "-LRB-"
 		elif senttok[i] == "-RRB-":
 			subt.label = "-RRB-"
-		if (senttok[i] not in string.punctuation + "-LRB-" + "-RRB-") and subt.label in string.punctuation + "-LRB-" + "-RRB-":
+		if (senttok[i] not in string.punctuation + "-LRB-" + "-RRB-" + "...") and subt.label in string.punctuation + "-LRB-" + "-RRB-" + "...":
 			print("match")
 			subt.label = 'N-Head'
+	# create three lists of equal lengths: one list non-punctuation token strings, one list of tuples prepending punctuation, and one list of tuples for appending punctuation
+	# iterate through the tree to create the three lists simultaneously
+	non_punct_tokens = []
+	prepunct_tokens = [[] for t in senttok if t not in string.punctuation + "-LRB-" + "-RRB-" + "..."]
+	postpunct_tokens = [[] for t in senttok if t not in string.punctuation + "-LRB-" + "-RRB-" + "..."]
+	token_counter = 0
+	initial_punct = {'-LRB-', '[', '{'}
+	trailing_punct = {'.', '...', ',', ';', '!', '?', '-RRB-', ']', '}', "'", '"', ':', '-'}
```

Comment:
Is this necessary, or can all punctuation not in `initial_punct` be assumed to be trailing?

Reply:
I removed this redundancy in https://github.com/nschneid/activedop/pull/47/commits/80285bf669d29dfb9550039f6940b06f10bce436